Change debug info default variable tracking system

BrianBohe opened the pull request:
Varible Scope info is being disabled. Variable
Live Range is being enabled. Debug info generated
for debug code is the same, changes only impact on
optimized code.

Variable Live Range updates variable's location
each time something happens to variable's
liveness. For more information see
https://github.com/dotnet/runtime/blob/master/docs/design/features/variabletracking.md

BrianBohe committed:
Change debug info default variable tracking system.

Varible Scope info is being disabled. Variable
Live Range is being enabled. Debug info generated
for debug code is the same, changes only impact on
optimized code.

Variable Live Range updates variable's location
each time something happens to variable's
liveness. For more information see
https://github.com/dotnet/runtime/blob/master/docs/design/features/variabletracking.md

BrianBohe commented:
I will open an issue to develop a baseline/contract/protocol so we can check no one downgrade/reduce/break debug info on optimize code with future changes.

BrianBohe commented:
/azp list

BrianBohe commented:
@BruceForstall

BrianBohe commented:
@dotnet/jit-contrib

BrianBohe commented:
@dotnet-bot Test runtime

BruceForstall commented:
/azp list

BrianBohe commented:
/azp run runtime

BrianBohe commented:
/azp run runtime

BrianBohe commented:
/azp run runtime

BrianBohe commented:
/azp run runtime

BrianBohe commented:
/azp list

BrianBohe commented:
/azp list

BrianBohe commented:
/azp list

BrianBohe commented:
/azp list

BrianBohe commented:
/azp list

BrianBohe commented:
/azp run runtime (Test crossgen-comparison Linux arm checked)

BrianBohe commented:
/azp run runtime-coreclr crossgen2 outerloop

BrianBohe commented:
/azp help

BrianBohe commented:
/azp run

BrianBohe commented:
/azp run runtime-coreclr crossgen2, runtime-coreclr crossgen2 outerloop

BrianBohe commented:
/azp run runtime

BrianBohe commented:
System.Private.CoreLive.dll increases from 9183232 bytes to 9367552 bytes, which is a 2% of growth with my changes.

BruceForstall commented:
@BrianBohe Is this a Release build of System.Private.CoreLib.dll? Can you collect the data for all the framework assemblies?

@tommcdon @noahfalk @jkotas @stephentoub This change is to generate more debug info for locals to generate better location info in optimized code. Is a 2% assembly size increase (assuming that also holds for all the framework assemblies) acceptable?

jkotas commented:
> a 2% assembly size increase (assuming that also holds for all the framework assemblies) acceptable?

It sounds reasonable to me.

Have you analyzed methods with largest growth to verify whether it is all expected?

jkotas commented:
>  generate better location info in optimized code

Having a few before/after VS screenshots that demonstrate how it is better would help demonstrate value of this change.

BruceForstall commented:
> Have you analyzed methods with largest growth to verify whether it is all expected?

@BrianBohe It would be good to do this analysis on S.P.C.dll and maybe one or two other large ones.

BrianBohe commented:
> @BrianBohe Is this a Release build of System.Private.CoreLib.dll? Can you collect the data for all the framework assemblies?
> 
> @tommcdon @noahfalk @jkotas @stephentoub This change is to generate more debug info for locals to generate better location info in optimized code. Is a 2% assembly size increase (assuming that also holds for all the framework assemblies) acceptable?

Yes, It is a Release build.

BrianBohe commented:
> > Have you analyzed methods with largest growth to verify whether it is all expected?
> 
> @BrianBohe It would be good to do this analysis on S.P.C.dll and maybe one or two other large ones.

Measured on a Release build, listing the only files that changed on artifacts\bin\coreclr\Windows_NT.x64.Release:

- System.Private.Corelib.dll increases 2%

- clrjit.dll and protononjit.dll decreases 0.01%

BrianBohe commented:
> > generate better location info in optimized code
> 
> Having a few before/after VS screenshots that demonstrate how it is better would help demonstrate value of this change.

I don't remember how to use my coreclr in my visual studio but I can show you the difference in the jitdump.

I am using the following code:
`using System;
using System.Collections.Generic;

namespace example2
{
    class Program
    {
        static void Main(string[] args)
        {
            Random r = new Random();
            int a = r.Next();
            
            // create one list with random numbers
            List<int> one_list = new List<int>();
            for (int i = 0; i < a; i++)
            {
                int value = a * r.next();
                value <<= 2;
                one_list.Add(i * r.Next() + value);
            }

            // create other list
            List<int> other_list = new List<int>();
            for (int i = 0; i < a*3; i++)
            {
                int value = i + a * r.next();
                value >>= 1;
                one_list.Add(r.Next() + value);
            }

            // multiply them
            List<int> prod = new List<int>();
            for (int i = 0; i < one_list.Count; i++)
            {
                for (int j = 0; j < other_list.Count; j++)
                {
                    prod.Add(one_list[i] * other_list[j]);
                }
            }

            int acum = 0;
            int pairs_acum = 0;
            for (int i = 0; i < prod.Count; i++)
            {
                acum += prod.Count;
                if (0 == i % 2)
                {
                    pairs_acum += prod[i];
                }
            }

            Console.WriteLine(acum);
            Console.WriteLine(pairs_acum);
            Console.WriteLine(one_list.Count);
            Console.WriteLine(other_list.Count);
        }
    }
}
`

Using flags COMPlus_tieredcompilation = 0 and COMPlus_jitdump=Main you get:


With Variable Scope Info (the one we are turning off):

```
*************** In genSetScopeInfo()
VarLocInfo count is 15
*************** Variable debug info
15 live ranges
  0(   UNKNOWN) : From 00000000h to 0000000Eh, in rcx
  1(   UNKNOWN) : From 00000053h to 000000DDh, in rsi
  2(   UNKNOWN) : From 00000053h to 000000DDh, in rdi
  3(   UNKNOWN) : From 00000053h to 000001C1h, in rbx
  4(   UNKNOWN) : From 000000ACh to 000001D1h, in rbp
  5(   UNKNOWN) : From 00000106h to 000001B2h, in rsi
  6(   UNKNOWN) : From 00000192h to 000001B2h, in rdi
  6(   UNKNOWN) : From 00000174h to 0000017Dh, in rdi
  7(   UNKNOWN) : From 00000174h to 0000019Eh, in r14
  7(   UNKNOWN) : From 000001A1h to 000001B9h, in r14
  8(   UNKNOWN) : From 00000053h to 00000080h, in rbp
 10(   UNKNOWN) : From 000000ACh to 000000D5h, in r14
 12(   UNKNOWN) : From 00000106h to 00000150h, in rdi
 13(   UNKNOWN) : From 00000116h to 0000013Fh, in r14
 14(   UNKNOWN) : From 00000174h to 000001A1h, in r15
```


With Variable Live Range:

```
*************** In genSetScopeInfo()
VarLocInfo count is 24
*************** Variable debug info
24 live ranges
  0(   UNKNOWN) : From 00000000h to 0000000Eh, in rcx
  1(   UNKNOWN) : From 00000028h to 000000DDh, in rsi
  2(   UNKNOWN) : From 00000033h to 000000DDh, in rdi
  3(   UNKNOWN) : From 0000004Dh to 000001C1h, in rbx
  4(   UNKNOWN) : From 000000A0h to 000001D1h, in rbp
  5(   UNKNOWN) : From 000000F7h to 000001B2h, in rsi
  6(   UNKNOWN) : From 00000161h to 0000017Dh, in rdi
  6(   UNKNOWN) : From 0000017Fh to 000001B2h, in rdi
  7(   UNKNOWN) : From 00000164h to 0000019Eh, in r14
  7(   UNKNOWN) : From 000001A1h to 000001B9h, in r14
  8(   UNKNOWN) : From 0000004Fh to 00000080h, in rbp
  8(   UNKNOWN) : From 00000082h to 00000086h, in rbp
  9(   UNKNOWN) : From 00000063h to 00000063h, in r14
  9(   UNKNOWN) : From 00000067h to 00000073h, in r14
 10(   UNKNOWN) : From 000000A3h to 000000D5h, in r14
 10(   UNKNOWN) : From 000000D8h to 000000DDh, in r14
 11(   UNKNOWN) : From 000000BCh to 000000BCh, in r12
 11(   UNKNOWN) : From 000000BFh to 000000C8h, in r12
 12(   UNKNOWN) : From 000000F9h to 00000150h, in rdi
 12(   UNKNOWN) : From 00000152h to 0000015Fh, in rdi
 13(   UNKNOWN) : From 00000109h to 0000013Fh, in r14
 13(   UNKNOWN) : From 00000142h to 00000150h, in r14
 14(   UNKNOWN) : From 00000167h to 000001A1h, in r15
 14(   UNKNOWN) : From 000001A4h to 000001B2h, in r15


////////////////////////////////////////
////////////////////////////////////////
PRINTING VARIABLE LIVE RANGES:
IL Var Num 1:
[rsi [28 , DD )]
IL Var Num 2:
[rdi [33 , DD )]
IL Var Num 3:
[rbx [4D , 1C1 )]
IL Var Num 4:
[rbp [A0 , 1D1 )]
IL Var Num 5:
[rsi [F7 , 1B2 )]
IL Var Num 6:
[rdi [161 , 17D )rdi [17F , 1B2 )]
IL Var Num 7:
[r14 [164 , 19E )r14 [1A1 , 1B9 )]
IL Var Num 8:
[rbp [4F , 80 )rbp [82 , 86 )]
IL Var Num 9:
[r14 [63 , 63 )r14 [67 , 73 )]
IL Var Num 10:
[r14 [A3 , D5 )r14 [D8 , DD )]
IL Var Num 11:
[r12 [BC , BC )r12 [BF , C8 )]
IL Var Num 12:
[rdi [F9 , 150 )rdi [152 , 15F )]
IL Var Num 13:
[r14 [109 , 13F )r14 [142 , 150 )]
IL Var Num 14:
[r15 [167 , 1A1 )r15 [1A4 , 1B2 )]
////////////////////////////////////////
////////////////////////////////////////
```

BrianBohe commented:
Variable 9 and 11 didn't get reported without my changes. Other variables are being reported with more location/liveness information.

BrianBohe commented:
I didn't to close this PR, I just want to rebase my master branch. Can I open it again?

jkotas commented:
I have compiled your example above, but I am getting different results. I have to tweak it little bit (`next` has wrong casing in a few spots) that suggests you might have run something else.

Here is the jitdump that I am getting: https://gist.githubusercontent.com/jkotas/477da728c4e317ae106df056d2707573/raw/4947b7e51e9bb98130ba514f75763732d4286cd5/jitdump

The interesting parts of the dump is:
```
 14(   UNKNOWN) : From 00000211h to 00000244h, in rdi
 14(   UNKNOWN) : From 00000246h to 0000024Bh, in rdi
```

```
IN0094: 000240 add      esi, dword ptr [rax+4*rdx+16]

G_M7693_IG25:        ; offs=000244H, size=0007H, bbWeight=4    PerfScore 6.00, gcrefRegs=00001028 {rbx rbp r12}, byrefRegs=00000000 {}, byref, isz

IN0095: 000244 inc      edi
IN0096: 000246 cmp      edi, r13d
IN0097: 000249 jl       SHORT G_M7693_IG22
```

Why are there two ranges reported for variable #14 in edi in this case? Can this be reported as one range from `00000211h` to `0000024Bh`? Doing range merging like this should reduce the size overhead that you paying for the better debug info.

BrianBohe commented:
So I have changed the example to be deterministic (its below) and I will show we screens of a Debug build, a Release build without my changes (Old) and a Release build with my changes (New) on example same lines.
```
    class Program
    {
        static void Main(string[] args)
        {
            Random r = new Random(0);
            int a = r.Next(0,3000);

            // create one list with random numbers
            List<int> one_list = new List<int>();
            for (int i = 0; i < a; i++)
            {
                int b = r.Next(0, 3000); ;
                int value = a * b;
                value <<= 2;
                one_list.Add(i * r.Next(0,3000) + value);
            }

            // create other list
            List<int> other_list = new List<int>();
            for (int i = 0; i < a * 3; i++)
            {
                int b = r.Next(0, 3000);
                int value = i + a * b;
                value >>= 1;
                other_list.Add(r.Next(0,3000) + value);
            }

            // multiply them
            List<int> prod = new List<int>();
            for (int i = 0; i < one_list.Count; i++)
            {
                for (int j = 0; j < other_list.Count; j++)
                {
                    prod.Add(one_list[i] * other_list[j]);
                }
            }

            int acum = 0;
            int pairs_acum = 0;
            for (int i = 0; i < prod.Count; i++)
            {
                acum += prod.Count;
                if (0 == i % 2)
                {
                    pairs_acum += prod[i];
                }
            }

            Console.WriteLine(acum);
            Console.WriteLine(pairs_acum);
            Console.WriteLine(one_list.Count);
            Console.WriteLine(other_list.Count);
        }
    }
```

Debug
![image](https://user-images.githubusercontent.com/9440186/73563837-255ada80-4413-11ea-8e3c-ed22df0d09d4.png)
Old
![image](https://user-images.githubusercontent.com/9440186/73563852-31469c80-4413-11ea-9dc4-2af7de5e572f.png)
New
![image](https://user-images.githubusercontent.com/9440186/73563864-3c013180-4413-11ea-9bb6-9e5d36032caf.png)

Vars "value" an "b" are lost in Old version
D
![image](https://user-images.githubusercontent.com/9440186/73564846-563c0f00-4415-11ea-91d8-e9bc2fc99c30.png)
O
![image](https://user-images.githubusercontent.com/9440186/73564871-5fc57700-4415-11ea-8698-894869712783.png)
N
![image](https://user-images.githubusercontent.com/9440186/73564883-648a2b00-4415-11ea-88e3-6dcfaeb3b3d4.png)

In the last loop after we update "acum" var, the old version can't see the new value and lost it:
D
![image](https://user-images.githubusercontent.com/9440186/73568984-5ab8f580-441e-11ea-8246-b2b0d0212000.png)
O
![image](https://user-images.githubusercontent.com/9440186/73569013-699fa800-441e-11ea-94a2-d50cb81ca088.png)
N
![image](https://user-images.githubusercontent.com/9440186/73569001-63a9c700-441e-11ea-9ed9-edcc26d08383.png)

There are cases where the old version overlap two variables locations during an interval intersection of native offsets (when one was about to die and another dying of being spilled), this is not an example of that case. I showed that example during a presentation a year ago.

BrianBohe commented:
> I have compiled your example above, but I am getting different results. I have to tweak it little bit (`next` has wrong casing in a few spots) that suggests you might have run something else.
> 
> Here is the jitdump that I am getting: https://gist.githubusercontent.com/jkotas/477da728c4e317ae106df056d2707573/raw/4947b7e51e9bb98130ba514f75763732d4286cd5/jitdump
> 
> The interesting parts of the dump is:
> 
> ```
>  14(   UNKNOWN) : From 00000211h to 00000244h, in rdi
>  14(   UNKNOWN) : From 00000246h to 0000024Bh, in rdi
> ```
> 
> ```
> IN0094: 000240 add      esi, dword ptr [rax+4*rdx+16]
> 
> G_M7693_IG25:        ; offs=000244H, size=0007H, bbWeight=4    PerfScore 6.00, gcrefRegs=00001028 {rbx rbp r12}, byrefRegs=00000000 {}, byref, isz
> 
> IN0095: 000244 inc      edi
> IN0096: 000246 cmp      edi, r13d
> IN0097: 000249 jl       SHORT G_M7693_IG22
> ```
> 
> Why are there two ranges reported for variable #14 in edi in this case? Can this be reported as one range from `00000211h` to `0000024Bh`? Doing range merging like this should reduce the size overhead that you paying for the better debug info.

I can see the same is happening with variables 8, 11 and 12. In all of the cases, the difference of the native offset is the size of the opcode.

The boundaries are set by the linear register scan when it decides that a variables becomes alive/dead/spill/unspill. the new variable tracking system is just consuming that information. I think we could add this post-processing task as issue. I am not sure whether there is another case or not. It should be check if the difference between two live ranges of a variable in the same location is the opcode been produced on that offset.

jkotas commented:
> I think we could add this post-processing task as issue. I am not sure whether there is another case or not.

How hard is this to fix this? It would be better to fix the simple things like this that are contributing to the size regression before enabling this.

BruceForstall commented:
In the example above, the two live ranges are disjoint, and are missing the 
```
IN0095: 000244 inc      edi
```
instruction. Why is that?

BrianBohe commented:
> > I think we could add this post-processing task as issue. I am not sure whether there is another case or not.
> 
> How hard is this to fix this? It would be better to fix the simple things like this that are contributing to the size regression before enabling this.

I am not sure how hard it is but I have another internship project right now. The idea of opening an issue was because maybe its a good first task to do in the repo and we have other contributors. If no one takes it once I finish my intern project, I can do it.

BrianBohe commented:
Hi, I already presented some scenarios cases where there are improves. I would like this PR to be merged. Do you think we need more reviewers? Would you like to continue this on an issue?
@BruceForstall

jkotas commented:
Yes, the few examples are nice, but the inspection of the debug into delta shows inefficiencies that look unexpected. Unexpected discrepancies like this often point to bugs and real problems in some cases.

I think we should understand and ideally fix there discrepancies before merging this. (If the codegen team believes that these discrepancies should not prevent merging, I am happy to be overruled.)

cshung commented:
What is our plan here? Are we fixing the issue and proceed to get it merged so that it is available for 5.0, or are we abandoning it?

BrianBohe commented:
> What is our plan here? Are we fixing the issue and proceed to get it merged so that it is available for 5.0, or are we abandoning it?

I have just talked with @BruceForstall about this. I will do some research around the specific case that was pointed above and post the results here. If results show there is no problem merging variable live ranges on that case (or others if I found more), I will add a post processing to merge them.

BrianBohe commented:
Using the last peace of code I provided I have reviewed the examples and I can confirm that my changes are not introducing a bug. I am correctly reporting variable location given the liveness flags set on nodes.

Some questions I would like to answer:

1. Why are we reporting the variable to not "be alive" during inc operations?
2. Are there any other cases in which the same happens?
3. Is it safe/correct to merge the Variable Live Ranges in those cases?


1) Using the last peace of code I gave, we found the same behavior is happening with V16:

IL Var Num 16:
[r15 [191 , 1CB )r15 [1CE , 1DC )]

In IN645 it says to be last used (and then set as becoming dead) and saved in t185, it is then incremented with t186 and the result (t187) is stored in the same register.

We started reporting it back when STORE_LCL_VAR assign t187 in IN0079.

````
=============== Generating BB12 [11A..12B) -> BB10 (cond), preds={BB10,BB11} succs={BB13,BB10} flags=0x00000004.420b0020: i label target gcsafe bwd LIR 
BB12 IN (6)={V04 V05 V16 V03 V06 V07} + ByrefExposed + GcHeap
     OUT(6)={V04 V05 V16 V03 V06 V07} + ByrefExposed + GcHeap

Recording Var Locations at start of BB12
  V04(rbp)  V05(rsi)  V16(r15)  V03(rbx)  V06(rdi)  V07(r14)
Liveness not changing: 00000000000900E8 {V03 V04 V05 V06 V07 V16}
							Live regs: 00000000 {} => 0000C0E8 {rbx rbp rsi rdi r14 r15}
							GC regs: 00000000 {} => 00000068 {rbx rbp rsi}
							Byref regs: (unchanged) 00000000 {}

      L_M27759_BB12:

      G_M27759_IG12:        ; offs=0001C6H, funclet=00, bbWeight=2   
Label: IG13, GCvars=0000000000000000 {}, gcrefRegs=00000068 {rbx rbp rsi}, byrefRegs=00000000 {}

Scope info: begin block BB12, IL range [11A..12B)
Added IP mapping: 0x011A STACK_EMPTY (G_M27759_IG13,ins#0,ofs#0) label
Generating: N645 (???,???) [000395] ------------                 IL_OFFSET void   IL offset: 0x11a REG NA
Generating: N647 (  1,  1) [000185] ------------       t185 =    LCL_VAR   int    V16 loc15        u:3 r15 (last use) REG r15 $246
Generating: N649 (  1,  1) [000186] -c----------       t186 =    CNS_INT   int    1 REG NA $45
                                                              /--*  t185   int    
                                                              +--*  t186   int    
Generating: N651 (  3,  3) [000187] ------------       t187 = *  ADD       int    REG r15 $21e
							V16 in reg r15 is becoming dead  [000185]
							Live regs: 0000C0E8 {rbx rbp rsi rdi r14 r15} => 000040E8 {rbx rbp rsi rdi r14}
							Live vars: {V03 V04 V05 V06 V07 V16} => {V03 V04 V05 V06 V07}
IN0079:        inc      r15d
                                                              /--*  t187   int    
Generating: N653 (  3,  3) [000189] DA----------              *  STORE_LCL_VAR int    V16 loc15        d:4 r15 REG r15
							V16 in reg r15 is becoming live  [000189]
							Live regs: 000040E8 {rbx rbp rsi rdi r14} => 0000C0E8 {rbx rbp rsi rdi r14 r15}
							Live vars: {V03 V04 V05 V06 V07} => {V03 V04 V05 V06 V07 V16}
Added IP mapping: 0x0120 STACK_EMPTY (G_M27759_IG13,ins#1,ofs#3)
Generating: N655 (???,???) [000396] ------------                 IL_OFFSET void   IL offset: 0x120 REG NA
Generating: N657 (  1,  1) [000169] ------------       t169 =    LCL_VAR   ref    V05 loc4         u:2 rsi REG rsi $187
                                                              /--*  t169   ref    
Generating: N659 (???,???) [000485] ------------       t485 = *  PUTARG_REG ref    REG rcx
IN007a:        mov      rcx, rsi
							GC regs: 00000068 {rbx rbp rsi} => 0000006A {rcx rbx rbp rsi}
Generating: N661 (  3, 10) [000486] -c----------       t486 =    CNS_INT(h) long   0x7ffeeb369958 ftn REG NA
                                                              /--*  t486   long   
Generating: N663 (  5, 12) [000487] -c----------       t487 = *  IND       long   REG NA
                                                              /--*  t485   ref    this in rcx
                                                              +--*  t487   long   control expr
Generating: N665 ( 15,  8) [000170] --CXG-------       t170 = *  CALL      int    System.Collections.Generic.List`1[Int32][System.Int32].get_Count REG rax $4c6
							GC regs: 0000006A {rcx rbx rbp rsi} => 00000068 {rbx rbp rsi}
							Call: GCvars=0000000000000000 {}, gcrefRegs=00000068 {rbx rbp rsi}, byrefRegs=00000000 {}
IN007b:        call     [System.Collections.Generic.List`1[Int32][System.Int32]:get_Count():int:this]
Generating: N667 (  1,  1) [000168] ------------       t168 =    LCL_VAR   int    V16 loc15        u:4 r15 REG r15 $21e
                                                              /--*  t170   int    
                                                              +--*  t168   int    
Generating: N669 ( 17, 10) [000171] J--XG--N----              *  GT        void   REG NA $21f
IN007c:        cmp      eax, r15d
Generating: N671 ( 19, 12) [000172] ---XG-------              *  JTRUE     void   REG NA
IN007d:        jg       SHORT L_M27759_BB10
////////////////////////////////////////
////////////////////////////////////////
Variable Live Range History Dump for Block 12 
IL Var Num 3:
[rbx [ (G_M27759_IG02,ins#16,ofs#73), NON_CLOSED_RANGE ]; ]
IL Var Num 4:
[rbp [ (G_M27759_IG04,ins#5,ofs#26), NON_CLOSED_RANGE ]; ]
IL Var Num 5:
[rsi [ (G_M27759_IG06,ins#5,ofs#26), NON_CLOSED_RANGE ]; ]
IL Var Num 6:
[rdi [ (G_M27759_IG11,ins#3,ofs#11), NON_CLOSED_RANGE ]; ]
IL Var Num 7:
[r14 [ (G_M27759_IG12,ins#4,ofs#15), NON_CLOSED_RANGE ]; ]
IL Var Num 16:
[r15 [ (G_M27759_IG10,ins#3,ofs#8), (G_M27759_IG13,ins#0,ofs#0) ]; r15 [ (G_M27759_IG13,ins#1,ofs#3), NON_CLOSED_RANGE ]; ]
////////////////////////////////////////
////////////////////////////////////////
End Generating code for Block 12 
````

The variable track system is taking the information to report the end / start of locations given the liveness flags.

The result in the following code :

````
G_M27759_IG13:        ; func=00, offs=0001CBH, size=0011H, gcrefRegs=00000068 {rbx rbp rsi}, byrefRegs=00000000 {}, byref, isz
IN0079: 0001CB 41FFC7               inc      r15d
gcrReg +[rcx]
IN007a: 0001CE 488BCE               mov      rcx, rsi
New gcrReg live regs=00000068 {rbx rbp rsi}
gcrReg -[rcx]
[14B22460] ptr arg pop  0
IN007b: 0001D1 FF1551242900         call     [System.Collections.Generic.List`1[Int32][System.Int32]:get_Count():int:this]
IN007c: 0001D7 413BC7               cmp      eax, r15d
IN007d: 0001DA 7FC2                 jg       SHORT G_M27759_IG11
````


2) A similar case is when adding in a register that contains a variable listed as last used before:

For example, with variable 7 in R14:

IL Var Num 7:
[r14 [18E , 1C8 )r14 [1CB , 1E3 )]

````
G_M27759_IG12:        ; offs=0001BCH, size=000FH, bbWeight=2    PerfScore 7.50, gcrefRegs=00000068 {rbx rbp rsi}, byrefRegs=00000000 {}, byref

IN0075: 0001BC mov      rcx, rsi
IN0076: 0001BF mov      edx, r15d
IN0077: 0001C2 call     [System.Collections.Generic.List`1[Int32][System.Int32]:get_Item(int):int:this]
IN0078: 0001C8 add      r14d, eax

G_M27759_IG13:        ; offs=0001CBH, size=0011H, bbWeight=4    PerfScore 19.00, gcrefRegs=00000068 {rbx rbp rsi}, byrefRegs=00000000 {}, byref, isz

IN0079: 0001CB inc      r15d
IN007a: 0001CE mov      rcx, rsi
IN007b: 0001D1 call     [System.Collections.Generic.List`1[Int32][System.Int32]:get_Count():int:this]
IN007c: 0001D7 cmp      eax, r15d
IN007d: 0001DA jg       SHORT G_M27759_IG11
````

3) If we were debugging the assembly lines, it would have been safe to merge the two ranges in both cases as the result of the inc (or add) would be reflected after passing the assembly instruction and the value of the variable before/after execution is correct.

BrianBohe commented:
I want to debug more to see why the last used despite being used after.

@CarolEidt Have you seen this before?

BrianBohe commented:
Getting deep in the case of V07

This happens on BB11, which looks like this:

Before rationalize
````
------------ BB11 [10C..11A), preds={BB10} succs={BB12}
***** BB11
STMT00042 (IL 0x10C...0x118)
N009 ( 18, 12) [000196] -ACXG---R---              *  ASG       int    $21c
N008 (  1,  1) [000195] D------N----              +--*  LCL_VAR   int    V07 loc6         d:6 $21c
N007 ( 18, 12) [000194] --CXG-------              \--*  ADD       int    $21c
N005 ( 16, 10) [000193] --CXG-------                 +--*  CALL      int    System.Collections.Generic.List`1[Int32][System.Int32].get_Item $4c3
N003 (  1,  1) [000191] ------------ this in rcx     |  +--*  LCL_VAR   ref    V05 loc4         u:2 $187
N004 (  1,  1) [000192] ------------ arg1 in rdx     |  \--*  LCL_VAR   int    V16 loc15        u:3 $246
N006 (  1,  1) [000190] ------------                 \--*  LCL_VAR   int    V07 loc6         u:4 (last use) $245

````
After rationalize
````
------------ BB11 [10C..11A), preds={BB10} succs={BB12}
               [000394] ------------                 IL_OFFSET void   IL offset: 0x10c
N003 (  1,  1) [000191] ------------       t191 =    LCL_VAR   ref    V05 loc4         u:2 $187
N004 (  1,  1) [000192] ------------       t192 =    LCL_VAR   int    V16 loc15        u:3 $246
                                                  /--*  t191   ref    this in rcx
                                                  +--*  t192   int    arg1 in rdx
N005 ( 16, 10) [000193] --CXG-------       t193 = *  CALL      int    System.Collections.Generic.List`1[Int32][System.Int32].get_Item $4c3
N006 (  1,  1) [000190] ------------       t190 =    LCL_VAR   int    V07 loc6         u:4 (last use) $245
                                                  /--*  t193   int    
                                                  +--*  t190   int    
N007 ( 18, 12) [000194] ---XG-------       t194 = *  ADD       int    $21c
                                                  /--*  t194   int    
N009 ( 18, 12) [000196] DA-XG-------              *  STORE_LCL_VAR int    V07 loc6         d:6
````

And has this bbLiveIn and bbLiveOut variables
````
BB11 IN (6)={V04 V05 V03 V16 V06 V07} + ByrefExposed + GcHeap
     OUT(6)={V04 V05 V03 V16 V06 V07} + ByrefExposed + GcHeap
````


When we get to fbInterBlockLocalVarLiveness() to fill liveness info within BB11:

- As the tree is being traversed the nodes where V07 appeared are discovered in this order:
    1.
    ````
      N008 (  1,  1) [000195] D------N----              *  LCL_VAR   int    V07 loc6
      Def V07,T20 at [000195] life 00000000001100E8 -> 00000000000100E8
    ````
    It is a def, V07 is on the set of variables alive (life set) as it is in bbLiveOut.
    The set of vars volatileVar and BB11->bbScope are empty, so V07 is not in the set of vars to keep alive and we remove it from life set. 
    2.
    ````
      N006 (  1,  1) [000190] ------------              *  LCL_VAR   int    V07 loc6
      Ref V07,T20] at [000190] life 00000000000100E8 -> 00000000001100E8
    ````
    It is not a def, V07 is no longer on life set. We added a death flag to node and add V07 to life set.


- If we would have found those nodes in the opposite order, the story would have been as follows:
    1.
    ````
      N006 (  1,  1) [000190] ------------              *  LCL_VAR   int    V07 loc6
      Ref V07,T20] at [000190] life 00000000000100E8 -> 00000000001100E8
    ````
    It is not a def. V07 is in the life set. The death flag is unset. Life set is not modified.
    2.
    ````
      N008 (  1,  1) [000195] D------N----              *  LCL_VAR   int    V07 loc6
      Def V07,T20 at [000195] life 00000000001100E8 -> 00000000000100E8
    ````
    V07 is removed from life set.

Which should be the expected behaviour?

BrianBohe commented:
Also this [this peace of code](https://github.com/dotnet/runtime/blob/a80e6d4c69fcfa1e17e17130c144a567b23a12a9/src/coreclr/src/jit/lsrabuild.cpp#L2272-L2288) sets the lastUse on the last refPosition of each variable alive in bbLiveOut to false. But the last ref in this interval is not N006.

BrianBohe commented:
Here is the raw dump of my main function [https://raw.githubusercontent.com/BrianBohe/runtime/efb8b12bfec7537e784198ee5da23c5e836aab45/main_dump.out](https://raw.githubusercontent.com/BrianBohe/runtime/efb8b12bfec7537e784198ee5da23c5e836aab45/main_dump.out)

BrianBohe commented:
@BruceForstall, How can we proceed ?

CarolEidt commented:
Regarding the liveness of V07 near the end of block BB11 in the above example:

First, we cannot traverse the liveness in the other order - the use at `[000190]` happens before the def, and the variable is live-out of the block after the def.

> I want to debug more to see why the last used is set despite being used after.

The use at `[000190] LCL_VAR v07` is a last use. That value is not used after.

The reason that it goes dead is that the `ADD` at `[000194]` is defining a new value, i.e. `ADD(V07,CALL)`. This value is defined into a "tree temp", which is then assigned to `V07`, at which point `V07` becomes live again. The awkward thing is that in the case where the value isn't moved, it could be considered to go live at the time that the result of the `ADD` is defined.

I suspect that it would not be a terribly difficult thing to work around this, but I'm not certain where the best place to do it would be. Perhaps at the time we open a new range at the `STORE_LCL_VAR` we could observe that 1) the value that's going live was defined by the previous instruction and 2) the previous live range for that variable ended at that instruction. Then we could merge the live ranges.

BrianBohe committed:
Updating VariableLiveRange

When a variable is being born at the same location it
died at the previous assembly instruction (in the same
group of instructions), we reported as it was alive
the whole time.

BrianBohe commented:
Given the fact that:
- the this is the expected behavior of the traversal and CodeGen
- I haven't found an assembly instruction that could lead us to wrong debug info during just that line.

My proposal is to extend the VariableLiveRange in the case when a variable is being born in the same location it was dying the assembly instruction before.

BrianBohe committed:
Fixing format problems

BrianBohe committed:
Adding comments

BrianBohe commented:
These were the previous variable ranges
```
IL Var Num 1:
[rsi [2A , 107 )]
IL Var Num 2:
[rdi [3D , 107 )]
IL Var Num 3:
[rbx [57 , 1EB )]
IL Var Num 4:
[rbp [BA , 1FB )]
IL Var Num 5:
[rsi [121 , 1DC )]
IL Var Num 6:
[rdi [18B , 1A7 )rdi [1A9 , 1DC )]
IL Var Num 7:
[r14 [18E , 1C8 )r14 [1CB , 1E3 )]
IL Var Num 8:
[rbp [59 , 9A )rbp [9C , A0 )]
IL Var Num 9:
[rax [6E , 6E )]
IL Var Num 10:
[r14 [75 , 75 )r14 [79 , 8D )]
IL Var Num 11:
[r14 [BD , FF )r14 [102 , 107 )]
IL Var Num 12:
[rax [D7 , D7 )]
IL Var Num 13:
[r12 [DE , DE )r12 [E1 , F2 )]
IL Var Num 14:
[rdi [123 , 17A )rdi [17C , 189 )]
IL Var Num 15:
[r14 [133 , 169 )r14 [16C , 17A )]
IL Var Num 16:
[r15 [191 , 1CB )r15 [1CE , 1DC )]
```

and these are new ones

```
IL Var Num 1:
[rsi [2A , 107 )]
IL Var Num 2:
[rdi [3D , 107 )]
IL Var Num 3:
[rbx [57 , 1EB )]
IL Var Num 4:
[rbp [BA , 1FB )]
IL Var Num 5:
[rsi [121 , 1DC )]
IL Var Num 6:
[rdi [18B , 1DC )]
IL Var Num 7:
[r14 [18E , 1E3 )]
IL Var Num 8:
[rbp [59 , A0 )]
IL Var Num 9:
[rax [6E , 6E )]
IL Var Num 10:
[r14 [75 , 8D )]
IL Var Num 11:
[r14 [BD , 107 )]
IL Var Num 12:
[rax [D7 , D7 )]
IL Var Num 13:
[r12 [DE , F2 )]
IL Var Num 14:
[rdi [123 , 189 )]
IL Var Num 15:
[r14 [133 , 17A )]
IL Var Num 16:
[r15 [191 , 1DC )]
```

[The complete raw output is here](https://raw.githubusercontent.com/BrianBohe/runtime/experiments_pr/post_processing.out)

jkotas commented on src/coreclr/src/jit/codegencommon.cpp:
```diff
@@ -11466,9 +11466,19 @@ void CodeGenInterface::VariableLiveKeeper::VariableLiveDescriptor::startLiveRang
+    if (!m_VariableLiveRanges->empty() && m_VariableLiveRanges->back().m_EndEmitLocation.IsPreviousIns(emit) &&
+        siVarLoc::Equals(&varLocation, &(m_VariableLiveRanges->back().m_VarLocation)))
+    {
+        // The variable is being born at the exactly same place just one assembly instructtion.
```
```suggestion
        // The variable is being born at the exactly same place just one assembly instruction.
```

CarolEidt replied:
I would rephrase this whole comment as:
```
// The variable is being born just after the instruction at which it died.
// In this case, i.e. an update of the variable's value, we coalesce the live ranges.
```
This is more accurate - it actually does die before being reborn, but because it's an update we can report the contiguous range.

CarolEidt commented on src/coreclr/src/jit/emit.cpp:
```diff
@@ -65,6 +65,21 @@ UNATIVE_OFFSET emitLocation::GetFuncletPrologOffset(emitter* emit) const
 
+//------------------------------------------------------------------------
+// IsPreviousIns: Returns true if the emitter is on the next instruction
+//  of the same group than this emitLocation.
```
I would suggest "as" instead of "than"

CarolEidt commented on src/coreclr/src/jit/emit.cpp:
```diff
@@ -65,6 +65,21 @@ UNATIVE_OFFSET emitLocation::GetFuncletPrologOffset(emitter* emit) const
+{
+    assert(Valid());
+    bool sameGroup  = ig == emit->emitCurIG;
+    bool sameInsNum = emitGetInsNumFromCodePos(codePos) == emitGetInsNumFromCodePos(emit->emitCurOffset()) - 1;
```
I would rename this as `isPreviousInsNum` as it's not actually the same.
And I would rename `sameGroup` as `isSameGroup` (`sameGroup` could mean the actual group number).

CarolEidt commented on src/coreclr/src/jit/emit.cpp:
```diff
@@ -65,6 +65,21 @@ UNATIVE_OFFSET emitLocation::GetFuncletPrologOffset(emitter* emit) const
+bool emitLocation::IsPreviousIns(const emitter* emit) const
+{
+    assert(Valid());
+    bool sameGroup  = ig == emit->emitCurIG;
```
Please put parentheses around the `==` expression.

CarolEidt reviewed (commented):
A few comments that should be changed, but otherwise LGTM

BrianBohe committed:
Renaming functions and updating comments 

BrianBohe commented:
Thanks @jkotas and @CarolEidt for the suggestions, I have applied them all. How do we proceed now?

jkotas commented:
> System.Private.CoreLib.dll increases from 9183232 bytes to 9367552 bytes, which is a 2% of growth with my changes.

What is this growth with these fixes?

LGTM otherwise.

BrianBohe commented:
> > System.Private.CoreLib.dll increases from 9183232 bytes to 9367552 bytes, which is a 2% of growth with my changes.
> 
> What is this growth with these fixes?
> 
> LGTM otherwise.
@jkotas, with a Release build:
- System.Private.CoreLib.dll increases from 9432064 to 9504256 (~ +0.76 %)
- linuxnonjit.dll increases from 1095680 to 1096192 (~ +0.04 %)
- superpmi-shim-collector.dll increases from 205312 to 205824 (~ +0.2 %)
- clrjit.dll decreases from 1213952 to 1211904 (~ -0.1 %)

BruceForstall commented:
@briansull Can you please review this from the JIT perspective?

BrianBohe commented:
/azp help

BrianBohe commented:
/azp run runtime

briansull commented:
I reviewed the older code (under `#ifdef USING_VARIABLE_LIVE_RANGE`)
as this change just enables the new method of recording variable live range debug info.

And I give my approval for checkin.

briansull reviewed (approved):
Looks Good.

BrianBohe commented:
@BruceForstall What should we do with the instances failing?

BruceForstall commented:
> @BruceForstall What should we do with the instances failing?

You mean: what about test failures? What kind of failures? It looks like you set them to rerun?

In general, I would rerun until they pass. Unless there are known problems such as documented in https://github.com/dotnet/runtime/issues/702.

BrianBohe commented:
> 
> 
> > @BruceForstall What should we do with the instances failing?
> 
> You mean: what about test failures? What kind of failures? It looks like you set them to rerun?
> 
> In general, I would rerun until they pass. Unless there are known problems such as documented in #702.

@BruceForstall 
I mean the two pipelines: runtime and runtime (Test crossgen-comparison Linux arm checked). Those have been failing since Friday. The logs showed cancellations due to timeouts, errors publishing artifacts that already exists, and I/O errors due to not finding a file. I don't find those cases listed on the issue you provided.

BruceForstall commented:
@BrianBohe Looks like only crossgen-comparison failed due to infra error. I'd go ahead and merge.
